### PR TITLE
[mkldnn-v1.0] Add MKL-DNN Convolution

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
add first mkldnn v1.0 OP: mkldnn conv; pass mnist training test; pass make lint

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here
@TaoLv @PatricZhao @ZhennanQin

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_base.cc [hunk 1]
```diff
@@ -530,13 +530,11 @@ bool MKLDNNStorageType(const nnvm::NodeAttrs &attrs,
     if (v == - 1) v = kDefaultStorage;
 
   DispatchMode wanted_mode;
-#if MXNET_USE_MKLDNN == 1
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_base.cc]: Why remove this?

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h [hunk 2]
```diff
@@ -20,18 +20,18 @@
 /*!
  * \file mkldnn_convolution-inl.h
  * \brief
-*/
+ */
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_convolution-inl.h]: Please revert the change.

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h [hunk 3]
```diff
@@ -20,18 +20,18 @@
...
-#if MXNET_USE_MKLDNN == 1
+#if MXNET_USE_MKLDNN == 100
 
-#include <vector>
 #include <utility>
+#include <vector>
 #include "../convolution-inl.h"
-#include "./mkldnn_ops-inl.h"
 #include "./mkldnn_base-inl.h"
+#include "./mkldnn_ops-inl.h"
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_convolution-inl.h]: Why change the order?

**Comment by pengzhao-intel** [comment]
@TaoLv @ciyongch @ZhennanQin please help to review :)

File: src/operator/nn/mkldnn/mkldnn_base-inl.h [hunk 4]
```diff
@@ -277,6 +277,19 @@ inline static mkldnn::memory::desc GetWeightDesc(const NDArray &arr,
...
 }
 
+inline static const std::vector<NDArray> GetMKLDNNInputArray(const std::vector<NDArray> &inputs) {
+  std::vector<NDArray> ret;
+  ret.reserve(inputs.size());
+  for (const auto &in : inputs) {
+    if (in.IsView() && in.IsMKLDNNData()) {
+      ret.push_back(in.Reorder2Default());
+    } else {
+      ret.push_back(in);
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_base-inl.h]: Return `inputs` here to avoid overhead?

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h [hunk 5]
```diff
@@ -79,54 +79,63 @@ struct MKLDNNConvFullParam {
...
+    return *bwd_weights_pd_;
+  }
+
+ private:
+  std::shared_ptr<mkldnn::convolution_backward_data::primitive_desc> bwd_data_pd_;
+  std::shared_ptr<mkldnn::convolution_backward_weights::primitive_desc> bwd_weights_pd_;
+  std::shared_ptr<mkldnn::convolution_backward_data> bwd_data_;
+  std::shared_ptr<mkldnn::convolution_backward_weights> bwd_weight_;
+};
+
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_convolution-inl.h]: Please use `weight` or `weights`  consistently.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 6]
```diff
@@ -57,7 +58,7 @@ mkldnn::convolution_forward::primitive_desc GetConvFwdImpl(const MKLDNNConvFullP
   auto bias_md =
       bias ? (param.mkldnn_param.quantized ? GetMemDesc(*bias, mshadow::kInt32) : GetMemDesc(*bias))
            : mkldnn::memory::desc{
-             {}, mkldnn::memory::data_type::data_undef, mkldnn::memory::format::any};
+                 {}, mkldnn::memory::data_type::undef, mkldnn::memory::format_tag::any};
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: indent

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 7]
```diff
@@ -126,13 +127,12 @@ mkldnn::convolution_forward::primitive_desc GetConvFwdImpl(const MKLDNNConvFullP
 
   if (param.conv_param.dilate.ndim() == 0 && bias_md_ptr == nullptr) {
     mkldnn::convolution_forward::desc desc(prop, mkldnn::algorithm::convolution_direct, data_md,
-                                           weight_md, out_md, strides, padding, padding,
-                                           mkldnn::padding_kind::zero);
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: Is `padding_kind`  disabled or `zero` becomes default option from MKL-DNN v1.0?

File: src/common/exec_utils.h [hunk 8]
```diff
@@ -59,15 +59,15 @@ inline bool SetupDefaultBlobsIn(const std::vector<NDArray>& src,
   for (size_t i = 0; i < src.size(); i++) {
     auto& nd = src[i];
     bool is_default = nd.storage_type() == kDefaultStorage;
-#if MXNET_USE_MKLDNN == 1
+#if MXNET_USE_MKLDNN == 100
```
Comment (ZhennanQin) [inline src/common/exec_utils.h]: If we upgrade mkldnn to version 1.1, shall we have to change this to something like 110? I prefer not encoding the version into this macro.

File: src/operator/operator_common.h [hunk 9]
```diff
@@ -526,17 +526,46 @@ class OpSignature {
...
+    auto desc = mem.get_desc();
+    hash = hash * 2 + desc.data.format_kind;
+    eles.push_back(desc.data.format_kind);
     hash = hash * 2 + desc.data.data_type;
     eles.push_back(desc.data.data_type);
     for (int i = 0; i < desc.data.ndims; i++) {
       hash = hash * 2 + desc.data.dims[i];
       eles.push_back(desc.data.dims[i]);
     }
+    switch (desc.data.format_kind) {
```
Comment (ZhennanQin) [inline src/operator/operator_common.h]: Hashing mkldnn memory is more heavy than before, so shall we cache the hash result somewhere to reduce the hash overhead?

File: src/common/exec_utils.h (hunk 8 above)
Comment (TaoLv) [reply src/common/exec_utils.h]: No. Here we simply use 100 and 1 to distinguish the code for MKL-DNN v0.x and v1.x so we don't need to remove v0.x integration code and hence convenient to review. When merging this branch back to master, we will change all `MXNET_USE_MKLDNN == 100` back to `MXNET_USE_MKLDNN == 1`.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h (hunk 4 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h]: pls let author @ZhennanQin help to check

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h (hunk 5 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_convolution-inl.h]: shall we all use weight or weights ? @ZhennanQin

File: src/operator/nn/mkldnn/mkldnn_convolution.cc (hunk 7 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_convolution.cc]: mkldnn_padding_kind_t  Not used anymore
https://intel.github.io/mkl-dnn/dev_guide_transition_to_v1.html

File: src/operator/nn/mkldnn/mkldnn_convolution.cc (hunk 6 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_convolution.cc]: done

File: src/operator/operator_common.h (hunk 9 above)
Comment (rongzha1) [reply src/operator/operator_common.h]: Seems need extra for ( loop: desc.data.ndims) && for(loop: desc.data.format_desc.blocking.inner_nblks), I will do some profile to decide whether hash rslt needs cached

Comment (ZhennanQin) [reply src/operator/operator_common.h]: OK. Ideally, we can save the hash result into MKLDNNMemory, only computing it when hash result doesn't exist.

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h (hunk 5 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_convolution-inl.h]: This is a difference between mkldnn and mxnet. mxnet prefer to use `weight` while mkldnn prefer `weights`. So I don't know which should be the boundary of mkldnn and mxnet. @rongzha1 You can make the decision.

File: src/operator/nn/mkldnn/mkldnn_base-inl.h (hunk 4 above)
Comment (ZhennanQin) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h]: Yes, we can do that.

File: src/operator/operator_common.h (hunk 9 above)
Comment (rongzha1) [reply src/operator/operator_common.h]: in tests/python/train/test_conv.py  
fwd cache cost: 2.603
total fwd cost: 505.974

bwd cache cost: 1.798
total bwd cost: 720.74

So maybe not that necessary to optimize this code ?

File: src/operator/nn/mkldnn/mkldnn_base-inl.h (hunk 4 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_base-inl.h]: done

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h (hunk 5 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_convolution-inl.h]: let's all use weight as mkldnn_fully_connect.cc did. So change bwd_weights_pd_ to bwd_weight_pd_

**Comment by pengzhao-intel** [comment]
@ZhennanQin @ciyongch @TaoLv are all concerns resolved? Please help take a review again or approve the PR?

File: src/operator/nn/convolution.cc [hunk 10]
```diff
@@ -60,8 +60,12 @@ static void ConvolutionComputeExCPU(const nnvm::NodeAttrs& attrs,
   const ConvolutionParam& params = nnvm::get<ConvolutionParam>(attrs.parsed);
   if (SupportMKLDNNConv(params, inputs[0])) {
     MKLDNN_OPCHECK_INIT(false, outputs.size(), inputs, outputs);
-    const auto mkldnn_inputs = GetMKLDNNInputArray(inputs);
-    MKLDNNConvolutionForward(attrs, ctx, mkldnn_inputs, req, outputs);
+    if (CheckMKLDNNInputArrayIsView(inputs)) {
```
Comment (ZhennanQin) [inline src/operator/nn/convolution.cc]: Make line 63 - 67 into a helper function, like
```
MKLDNNRun(MKLDNNConvolutionForward, attrs, ctx, inputs, req, outputs);
```
Then we can use this for all operators.

**Comment by pengzhao-intel** [comment]
merging now.
